Automattic/jetpack
# IDC: URL secret to detect multi-domain sites

## Proposed changes:
* Introduce URL secret to detect multi-domain sites.

### Other information:


## Jetpack product discussion
p9dueE-7BD-p2#comment-9897

## Does this pull request change what data or activity we track or use?
No.

## Testing instructions:
1. Install the "Jetpack Debug" plugin, activate "IDC Simulation" and "REST API Tester".
2. Connect Jetpack.
3. Go to "Jetpack Debug -> IDC Simulation", enter a Spoof URL and enable "IDC Simulation". Click "Send Remote Request".
4. You should see the IDC screen. Run `wp option get jetpack_options` and confirm that it contains the URL Secret value:
```
'identity_crisis_url_secret' =>
  array (
    'secret' => 'random_string',
    'expires_at' => 1690162076,
  ),
```
5. Go to "Jetpack Debug -> "REST API Tester" and send a GET request to `jetpack/v4/identity-crisis/fetch-url-secret`. Confirm that you get the 403 error:
```
{
    "code": "invalid_user_permission_identity_crisis",
    "message": "You do not have the correct user permissions to perform this action.",
    "data": {
        "status": 403
    }
}
```
6. Use 3122a-pb in WPCOM sandbox via `wpsh` to test the `jetpack.idcUrlValidation` and `/jetpack/v4/identity-crisis/fetch-url-secret` endpoints.
Confirm that the secret in both API responses matches, and expiration timestamp (only returned by the "fetch" endpoint) is approximately 5 minutes in the future.
